### PR TITLE
Cluster Operator Config and Unit Test

### PR DESCRIPTION
**Pull request by jeizadi**
Add Config to the Cluster Operator based on viper package.

### CODE REVIEW

File: cmd/manager/config.go
Comment (hryan-ib) [inline cmd/manager/config.go:13]: This variable wasn't used until now, so I will have to update the helm chart to include this env variable. This is me just making a mental note.

File: kops/kops.go
Comment (hryan-ib) [inline kops/kops.go]: kops.go in master is no longer using the docker approach for kops and is instead using a binary. This makes the command structure a bit different, so when you merge with master all kops function command strings can be updated to call just the binary and not use any of this docker stuff!

File: kops/kops_test.go
Comment (hryan-ib) [inline kops/kops_test.go]: these test cases can also be updated when kops.go from master is merged since the command string will look a bit different

File: utils/envs.go
Comment (hryan-ib) [inline utils/envs.go:45]: Currently in master, the kops commands built in kops.go no longer use this function to grab the environment variables. The binary runs because the values are exported in the makefile or, if running on a cluster, when the deployment is created. The variables that come through viper either need to be exported to make it work like it is in master right now (or maybe they are already being exported like you have it), or we need to add this function into the binary command so that the variables are there.

File: kops/kops.go
Comment (hryan-ib) [inline kops/kops.go:15]: Merging from master will add a new 'path' element of the struct which allows for the binary to run from the defined path. That may be a variable you want to include in the Viper setup

File: kops/kops.go
Comment (hryan-ib) [inline kops/kops.go:18]: Currently in master, this is hanging useless since the binary approach has taken over (I forgot to take it out) so feel free to kill it when you merge master in and stick to a function that sets the variables at runtime of the command

File: cmd/manager/main.go
Comment (hryan-ib) [inline cmd/manager/main.go]: When the operator is ran in a real kubernetes cluster, the AWS access key and access key ID will not be exported with it as it using existing keys on the cluster. I think this check, and the one for the key ID, would fail when starting up while running as a container so we would need to check the development variable to see if these should be set or not

**Review by hryan-ib — CHANGES_REQUESTED** [review]
The main work is looking good to me, here are some things to address (see comments for more info):
- Could you provide a little demo of running the operator with your changes implemented? Is there extra inputs we need to provide when running the start up command or will running this look the same as before? If it's different, then we will need to update the makefile to change the command, and possibly the Dockerfile.
- Merge master into your fork to get all the current changes. You will probably need to overhaul the kops functions a little bit to get it working with the binary approach that master is using.
- We aren't passing in the AWS credentials as variables when running this as a container, so the checks for those values would fail when we put it on a cluster. We either need to choose to ignore these checks when we are not in development stage, or find a different way to check for the creds when running as a container.

Please let me know if you have any questions or need some help, and please correct me if any of my comments are misunderstanding your changes!

File: cmd/manager/main.go
Comment (hryan-ib) [inline cmd/manager/main.go:68]: Another thing, I also added a "reap" value into a ReconcilerConfig that would take the REAPER environmental variable and instantiate the reconciler with that value. You may want to incorporate that value into this setup when you merge

File: cmd/manager/main.go
Comment (jeizadi) [reply cmd/manager/main.go]: We need some way to check whether the AWS secret keys are present when we set-up the cluster. How do we do that?

Comment (hryan-ib) [reply cmd/manager/main.go]: I think we could look at the ClusterOperatorDevelopment flag first to determine if we are running locally or not If thats true, then we want to check for the AWS creds and export them as you're doing. If not, then we can ignore the check and assume we are running on a cluster with a role. When on a cluster, there won't be explicit AWS creds exported to the container so I don't think we can check that properly. If there was no role on the cluster or it didn't have sufficient permissions, that would be an error we expect to see when deploying and would be resolved on the cluster side, not the application side

File: kops/kops.go
Comment (hryan-ib) [inline kops/kops.go:276]: making a note to myself that I need to change that to use the binary!

File: cmd/manager/config.go
Comment (hryan-ib) [inline cmd/manager/config.go]: remove the '/' from the default so it matches what the kops functions are expecting
```suggestion
	defaultKopsKubeDir = "kube"
```

File: kops/kops.go
Comment (hryan-ib) [inline kops/kops.go]: make `" --kubeconfig=" + viper.GetString("tmp.dir") + "/config-" + cluster.Name` to match what we pass to the function below
```suggestion
		" --kubeconfig=" +  viper.GetString("tmp.dir") + "/config-" + cluster.Name
```

File: kops/kops.go
Comment (hryan-ib) [inline kops/kops.go]: change to `file, err := ioutil.ReadFile(viper.GetString("tmp.dir") + "/config-" + cluster.Name)` which adds the extra slash in front of 'config' to make the file path correct
```suggestion
	file, err := ioutil.ReadFile(viper.GetString("tmp.dir") + "/config-" + cluster.Name)
```

**Review by hryan-ib — CHANGES_REQUESTED** [review]
Just a couple small syntax things to tidy up, but otherwise it looks great. I tested it out and with the changes I mentioned and my updates to the deployment and image, the operator seems to be functioning properly. I have PR #51 open that was meant to update the deploy for Shaun's PR but I will include some changes in the chart and Dockerfile that will allow us to properly run the controller with these changes. It can be merged once we merge this in.

File: kops/kops.go
Comment (hryan-ib) [inline kops/kops.go]:
```suggestion
		" -f ." + viper.GetString("kops.kube.dir") + "/" + tempConfigFile +
```
this fixes the path

File: kops/kops.go
Comment (hryan-ib) [reply kops/kops.go:276]: resolved in #51

File: cmd/manager/config.go
Comment (hryan-ib) [reply cmd/manager/config.go:13]: Resolved in #51

**Review by hryan-ib — APPROVED** [review]
LGTM

**Comment by hryan-ib** [comment]
@jeizadi I approved the PR but the checks won't run because we are trying to merge a fork of the repo. I think you need to go to your fork, settings, actions, and enable them so the checks will run on your end. Or we can give you write access to the repo if you want to move these changes to a branch

**Comment by hryan-ib** [comment]
These changes have been merged as a branch in another PR

**Pull request closed without merging**
